packages/lesswrong/client/AppGenerator.tsx

```diff
@@ -11,19 +11,29 @@ import { CookiesProvider } from 'react-cookie';
 import { BrowserRouter } from 'react-router-dom';
 import { ABTestGroupsUsedContext, RelevantTestGroupAllocation } from '../lib/abTestImpl';
 import type { AbstractThemeOptions } from '../themes/themeNames';
-import type { TimeOverride } from '../lib/utils/timeUtil';
+import type { SSRMetadata, TimeOverride } from '../lib/utils/timeUtil';
 import { LayoutOptionsContextProvider } from '../components/hooks/useLayoutOptions';
 
 // Client-side wrapper around the app. There's another AppGenerator which is
 // the server-side version, which differs in how it sets up the wrappers for
 // routing and cookies and such.
-const AppGenerator = ({ apolloClient, foreignApolloClient, abTestGroupsUsed, themeOptions, timeOverride }: {
+const AppGenerator = ({ apolloClient, foreignApolloClient, abTestGroupsUsed, themeOptions, ssrMetadata }: {
   apolloClient: ApolloClient<NormalizedCacheObject>,
   foreignApolloClient: ApolloClient<NormalizedCacheObject>,
   abTestGroupsUsed: RelevantTestGroupAllocation,
   themeOptions: AbstractThemeOptions,
-  timeOverride: TimeOverride,
+  ssrMetadata: SSRMetadata,
 }) => {
+  const [timeOverride, setTimeOverride] = useState<TimeOverride | null>({
+    currentTime: new Date(ssrMetadata.renderedAt),
+    cacheFriendly: ssrMetadata.cacheFriendly,
+    timezone: ssrMetadata.timezone
+  });
+
+  useEffect(() => {
+    setTimeOverride(null);
+  }, []);
+
```

Comment:
The old behaviour here was:
- The first (hydration) render would use the `renderedAt` of the SSR as the current time, which might be up to 90s behind due to on-server caching. This override would then be cleared without triggering a new render
- `timezone` is part of the cache key with on-server caching, so there could only be a mismatch if the user has _no_ timezone cookie to begin with

The new behaviour is:
- (CDN caching will not use the `timezone` as part of the key, as this would reduce the hit rate too much)
- The first render uses the `renderedAt` and `timezone` of the server so the hydration matches (but times may be off due to the timezone). This override this then cleared, triggering a second render
- There is also a `cacheFriendly` flag in the override which causes relative dates to be rendered as absolute (so if someone is on a slow connection or has javascript disabled, the first render will still make sense)

The fact that it always triggers a second render isn't ideal, but I checked and it seems like this is fairly fast (~50ms for a post with a large body on dev), presumably a lot of things are memoized on the first pass